From 1cac153be4b71e97ec61c894c3681e803a17f985 Mon Sep 17 00:00:00 2001
From: Cody P Schafer <dev@codyps.com>
Date: Tue, 18 Nov 2014 14:52:56 -0500
Subject: [PATCH 26/29] mk: always use bin as bindir for stage0 CFG_BUILD and
 pass it an -L to the TLIB dir

We can't trust stage0 rustc to actually know where our libdir is, so we
need to pass it the location explicitly.

bin is always used as the bindir because that is where it is extracted
to by the downloader (could be changed in the future if we're inclined)
---
 mk/main.mk | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/mk/main.mk b/mk/main.mk
index 0fb6f5f..2e6c6c8 100644
--- a/mk/main.mk
+++ b/mk/main.mk
@@ -333,7 +333,15 @@ define SREQ
 
 # Destinations of artifacts for the host compiler
 HROOT$(1)_H_$(3) = $(3)/stage$(1)
+
+ifeq ($(1)-$(3),0-$$(CFG_BUILD))
+# stage0 relative paths are fixed so we can bootstrap from snapshots
+# (downloaded snapshots drop their rustc in HROOT/bin)
+# libdir discrepancy is worked around with RUSTFLAGS below.
+HBIN$(1)_H_$(3) = $$(HROOT$(1)_H_$(3))/bin
+else
 HBIN$(1)_H_$(3) = $$(HROOT$(1)_H_$(3))/$$(CFG_BINDIR_RELATIVE)
+endif
 HLIB$(1)_H_$(3) = $$(HROOT$(1)_H_$(3))/$$(CFG_LIBDIR_RELATIVE)
 
 # Destinations of artifacts for target architectures
@@ -341,6 +349,15 @@ TROOT$(1)_T_$(2)_H_$(3) = $$(HLIB$(1)_H_$(3))/rustlib/$(2)
 TBIN$(1)_T_$(2)_H_$(3) = $$(TROOT$(1)_T_$(2)_H_$(3))/bin
 TLIB$(1)_T_$(2)_H_$(3) = $$(TROOT$(1)_T_$(2)_H_$(3))/lib
 
+# Don't trust stage0, be explicit about libraries
+# TODO: rather than specifying sysroot, we really want to tell which libdir to
+#       use (ie: the dir containing 'rustlib'). This would allow us to avoid
+#       passing the '-L' options.
+ifeq ($(1),0)
+RUSTFLAGS_S_$(1)_T_$(2)_H_$(3) += --sysroot "$$(HROOT$(1)_H_$(3))" \
+				  -L "$$(TLIB$(1)_T_$(2)_H_$(3))"
+endif
+
 # Preqrequisites for using the stageN compiler
 ifeq ($(1),0)
 HSREQ$(1)_H_$(3) = $$(HBIN$(1)_H_$(3))/rustc$$(X_$(3))
@@ -450,6 +467,7 @@ STAGE$(1)_T_$(2)_H_$(3) := \
 		$$(HBIN$(1)_H_$(3))/rustc$$(X_$(3)) \
 		--cfg $$(CFGFLAG$(1)_T_$(2)_H_$(3)) \
 		$$(CFG_RUSTC_FLAGS) $$(EXTRAFLAGS_STAGE$(1)) --target=$(2)) \
+		$$(RUSTFLAGS_S_$(1)_T_$(2)_H_$(3)) \
                 $$(RUSTC_FLAGS_$(2))
 
 PERF_STAGE$(1)_T_$(2)_H_$(3) := \
@@ -458,6 +476,7 @@ PERF_STAGE$(1)_T_$(2)_H_$(3) := \
 		$$(HBIN$(1)_H_$(3))/rustc$$(X_$(3)) \
 		--cfg $$(CFGFLAG$(1)_T_$(2)_H_$(3)) \
 		$$(CFG_RUSTC_FLAGS) $$(EXTRAFLAGS_STAGE$(1)) --target=$(2)) \
+		$$(RUSTFLAGS_S_$(1)_T_$(2)_H_$(3)) \
                 $$(RUSTC_FLAGS_$(2))
 
 endef
-- 
2.1.3

